### ISSUE TITLE
AssertionError in Mailbox.recv() causes context thread to stay running

### ISSUE THREAD

**Issue by courtarro** [zeromq/jeromq#179]
I am running a complex unit test locally that involves 5 nodes with various PUB/SUB and REQ/RES pairs. They're all talking via the public IP of the computer, but everything is on the same machine, in the same JVM.

Each endpoint has its own ZMQ.Context because it's a simulation of nodes that could be put on different computers across a network.

I'm running multithreaded. Each thread has its own ZMQ.Context object, along with its own socket (in any of the modes listed above). At the end of the test, I try to shut everything down gracefully, but I get an intermittent AssertionError:

```
Exception in thread "MessageSubscriber: CharlieOne" java.lang.AssertionError
    at zmq.Mailbox.recv(Mailbox.java:114)
    at zmq.SocketBase.process_commands(SocketBase.java:793)
    at zmq.SocketBase.recv(SocketBase.java:714)
    at org.zeromq.ZMQ$Socket.recv(ZMQ.java:1247)
    at org.zeromq.ZMQ$Socket.recv(ZMQ.java:1235)
    at MessageSubscriber$ListenerThread.run(MessageSubscriber.java:131)
```

"MessageSubscriber: CharlieOne" is the test thread that contains a ZMQ subscriber. The code that causes the AssertionError is [here](https://github.com/zeromq/jeromq/blob/master/src/main/java/zmq/Mailbox.java#L113).

Here is the code that runs in the class that owns that thread, to shut down the thread:

```
private ZMQ.Context context = ZMQ.context(1);
private ZMQ.Socket publisher = context.socket(ZMQ.PUB);

...

// Shut down the thread
listenerThread.run = false;
subscriber.close();
context.term();
try {
    listenerThread.join();
} catch (InterruptedException ex) {
    logger.warn("InterruptedException in thread.join()");
}
```

After the AssertionError occurs, `context.term()` blocks indefinitely, perhaps because the socket never really closed. I've tried catching the AssertionError, but that doesn't help. The `term()` call still blocks indefinitely.

The problem doesn't always happen. Sometimes the whole test runs, all the threads shut down, and the program exits. Most of the time, however, at least one of the threads blocks because of this assertion.

**Comment by trequartista10** [comment]
We are still seeing similar issues in our environment. Any fix for this ?

**Comment by daveyarwood** [comment]
@courtarro In the interests of resolving this issue: are you able to retry your code and see if the issue is still present with the latest version of JeroMQ?

As far as I know, `socket.recv()` should throw a ZMQException with an ETERM code if the socket is closed, or at least if the context is terminated. I'm not sure what the behavior is / should be if the socket is closed while blocked on a receive... would be interested to know more.

**Comment by courtarro** [comment]
Unfortunately I'm too far removed from the code I was editing when I originally contributed this bug report, so I can't really try again. Perhaps @trequartista10 can take over if he/she is still having the problem.

**Comment by daveyarwood** [comment]
It is unclear if this is still an issue in the latest version of JeroMQ. Closing for now -- if anyone observes this problem on the latest version, please open a new issue.
